### ISSUE TITLE
[JAVA] Support Javax injection for creation of API delegate instead of static factory pattern

### ISSUE THREAD

**Issue by JLLeitschuh** [swagger-api/swagger-codegen#3930]
##### Description

**Replace/Complement static factory with DI injection**

Currently, the generator for jersey and rest-easy use the static factory pattern to create the delegate that gets called by the REST API.

##### Swagger-codegen version

Whatever version this [gradle plugin](https://github.com/thebignet/swagger-codegen-gradle-plugin) in using, (I think it's the current release).

##### Swagger declaration file content or url

Use the pet-store as an example.

##### Command line used for generation

```groovy
swagger {
    inputSpec = "${project.projectDir.path}/spec/swagger.yaml"

    output = 'build/swagger'
    language = 'jaxrs-resteasy'

    additionalProperties = [
            'invokerPackage'   : 'edu.wpi.grip.web.swagger',
            'modelPackage'     : 'edu.wpi.grip.web.swagger.model',
            'apiPackage'       : 'edu.wpi.grip.web.swagger.api',
            'serializableModel': 'true'
    ]

    apis = ''
    models = ''
    supportingFiles = ''
}
```

##### Suggest a Fix

I'm using a custom `api.mustache` file that replaces the delegate creation with constructor injection.

```java
import javax.inject.Inject;

public class {{classname}}  {
    private final {{classname}}Service delegate;

    @Inject
    {{classname}}({{classname}}Service delegate) {
        this.delegate = delegate;
    }
```

If a zero argument constructor is needed for other reasons and other frameworks both solutions could be implemented in parallel.

```java
import javax.inject.Inject;

public class {{classname}}  {
    private final {{classname}}Service delegate;

    {{classname}}() { // Zero argument constructor because some frameworks need this
        this({{classname}}ServiceFactory.get{{classname}}(););
    }

    @Inject
    {{classname}}({{classname}}Service delegate) {
        this.delegate = delegate;
    }
```

**Comment by JLLeitschuh** [comment]
Is this something that would be pull request worthy?
Would this be too much of a breaking change because it would create a direct dependency upon `javax.inject` annotations?

This could also be done without the injection as you can use DI frameworks to bind to constructors.

**Comment by willgorman** [comment]
I'd definitely like to have something like this too.  Modifying the generated API class to add DI after the fact means giving up the ability to easily regenerate when the API spec is updated.

**Comment by JLLeitschuh** [comment]
If you are looking for my version of the template with dependency injection it can be found here.

**Comment by wing328** [comment]
@JLLeitschuh please kindly submit a PR so that we can review the enhancement.

**Comment by karnik** [comment]
Hi, I've also overwritten the templates to replace the static factory with DI. Since this could break something when CDI (JSF 299) is missing the introduction of a new language might be a good solution?

**jaxrs-resteasy-cdi** for example.

**Comment by JLLeitschuh** [comment]
I did something similar with Spring in such a way that it shouldn't break anything:
https://github.com/swagger-api/swagger-codegen/pull/4439

**Comment by karnik** [comment]
OK, adding a parameter should also work.

**Comment by mamghari** [comment]
@JLLeitschuh @wing328 we should be careful when using DI because it mainly depends on the DI framework you would like to use : Spring ? Guice ? HK2 ? etc...
So the swagger code generator should generate generic code and should be linked to any kind of Implementation framework. Using javax.inject is nice because it is the specification, then it leads to developers to choose the right implementation framework to use.

**Comment by JLLeitschuh** [comment]
I agree with using `javax.inject` I use `javax.inject` myself. The thing is, the code generator is generating code for specific frameworks. The spring generator is generating for spring. There are other generators generating for other frameworks. I don't know if you can decouple that well.

**Comment by xmeng1** [comment]
I think maybe we should decouple the interface and implement firstly, which is easier to mock implement for testing on clients. (I am using Jersey2, if I want to test the client code with invoking api, I need to run the server or mock the server. If the interface is decoupled, I can unit test with mock implementation.) Am I right to this point?

**Comment by JLLeitschuh** [comment]
That's the general idea, yes.

**Comment by xmeng1** [comment]
If we just make interface template with XXApi class, with only method signature. I think it should be easy and no affection for current usage. But is there anything else should be considered?

**Comment by JLLeitschuh** [comment]
You can see how I did this for spring as a part of #4439
I did this in a project that I was working on so that it would support DI this way:
https://github.com/WPIRoboticsProjects/GRIP/pull/676/files#diff-5c3416eb141bfe6552f621599c2a7474
(the link takes a second to load because it's a huge diff).

**Comment by xmeng1** [comment]
I am not very familiar with Spring MVC, I am familiar with Guice. If the interface level code is generated for the current XXXApi class, I think most of DI framework can use it easily. 
Actually, the urgent requirement for me is the unit test for functions that use the XXXApi. Currently, I use the constructor to inject the instance and mock the XXXApi totally in the unit test.

**Comment by wing328** [comment]
@xmeng1 we would love to see a [Guice](https://github.com/google/guice) generator as well 😃

**Comment by JLLeitschuh** [comment]
The example that I had above was using Guice. I don't remember what generator I used to make that work though. JavaRX maybe?
